feat: Adds navigationDelegate for handling links

Alexander-Ordina opened the pull request:
Adds NavigationDelegate to the chatbox for handling links.  

See https://github.com/talkjs/talkjs-flutter/issues/14

mrcnkoba commented:
Thank you @Alexander-Ordina for your contribution to our plugin. We will review your PR and get it released.

eteeselink commented:
Hi @Alexander-Ordina, first off, thanks for contributing this PR!

We're reviewing it, and while the code change is tiny, it may impact architecture changes we got lined up in the near future, so we'll need some time to evaluate it.  (note that we'll also need a signed CLA, even for such tiny code changes unfortunately, as discussed in the other PR you made)

Can you confirm that for your own project, you're unblocked since you can continue the forked version of the SDK until we merge this PR?

Alexander-Ordina commented:
Hi @eteeselink, I just got confirmation that the CLA is signed and send to your legal department. I'm guessing the architecture change you are talking about is the switch to InAppWebView, to get the same behavior with InAppWebView the [shouldOverrideUrlLoading ](https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/InAppWebView/shouldOverrideUrlLoading.html)property can be used. 

We do not plan to use the fork I made, the fork was only to make is possible to create a Pull Request, so until this is solved we are still blocked for rolling out Talkjs in our App.

bugnano commented:
> Hi @eteeselink, I just got confirmation that the CLA is signed and send to your legal department. I'm guessing the architecture change you are talking about is the switch to InAppWebView, to get the same behavior with InAppWebView the [shouldOverrideUrlLoading ](https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/InAppWebView/shouldOverrideUrlLoading.html)property can be used.
> 
You're totally right about the architecture change being the switch to InAppWebView.
Thank you for pointing us at the [shouldOverrideUrlLoading ](https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/InAppWebView/shouldOverrideUrlLoading.html)property, and as you can see its interface is slightly different than the NavigationDelegate.
Given that we would like to be able to switch implementation any time we want, can you come up with a solution that exposes the properties that you need in a WebView-implementation independent way for this PR, please?

Alexander-Ordina commented:
@bugnano I have added a onUrlNavigation function on the chatbox widget, this property can be used for adding the navigation logic independent of which webview is used. It is still needed to have a webview dependent implementation to translate this function for the type of webview used, for this I added the anonymous function _shouldNavigateToUrl which is used as the NavigationDelegate for the flutter webview. When switching to the InAppWebview this function will have to be rewritten to work with it.

Alexander-Ordina commented:
Hi @eteeselink, I understand that it might be surprising that we do not want to use the temporary fork. We want to prevent creating unnecessary forks which we have to maintain, we only do this if there is no other option. We try to make sure that we have plenty of time to solve issues with plugins before we release new features either by filing issues or creating pull requests. 

We expect that these issues can probably be solved before we go live with chat feature in our app, to make sure that is the case we would like to know when there is a new release for TalksJs plugin planned?

bugnano reviewed (commented):
Good work, I like it, but before merging this PR, I'd like to have the `UrlNavigationRequest` class to work with Dart's Uri regarding the `url` parameter, and to be explicit about the type.

Alexander-Ordina commented on lib/src/chatbox.dart:
```diff
@@ -52,6 +53,13 @@ class MessageActionEvent {
+enum UrlNavigationAction { allow, deny }
+
+class UrlNavigationRequest {
+  final url;
```
I Updated the UrlNavigationRequest it now had 3 properties:
```
final Uri? url;
final String rawUrl;
final bool isValidUrl;
```

This is to make sure that if the url is not a valid uri that it still is possible to pass the UrlNavigationRequest without throwing exceptions. 

I'm also thinking that it might be better to just pass the url as a String, this way any error handling or conversion to Uri Objects can be handled by the specific app using the plugin. 
What do you feel is best?

bugnano replied:
OK, so in this case let's use String for the url, and the validation is done by the user's code.
I thought that the Uri type could be the more appropriate one, but given that there could be some cases where the url is not a valid Uri, then my previous assumption was wrong.

bugnano commented on lib/src/chatbox.dart:
```diff
@@ -52,6 +53,30 @@ class MessageActionEvent {
+
+class UrlNavigationRequest {
+  final Uri? url;
+  final String rawUrl;
```
I would not expose the variables rawUrl and isValidUrl, because everything that's public needs to be supported forever, so the less we expose, the better.

Alexander-Ordina replied:
The problem is that the Uri.tryParse() method returns Null for an invalid url. If the rawUrl is not exposed there is no way for the app that uses plugin to figure out what the  original Url was for the link.  

It could be possible that a link is send with an invalid url, but in the app we still want to do something with this link. For example:
- We want to show an errormessage containing the url
- We could have a special url that opens a page in the app, maybe a url like: app://about would navigate to the app's about page. This is not a valid uri, but could be wanted behavior in a App. 

I think it would best to make the type for url a String this way the plugin stays simple and the rawUrl and isValidUrl can be removed.

```
class UrlNavigationRequest {
    final String url;
   ...
```

bugnano replied:
Yes, let's keep things simple and use String.

Alexander-Ordina replied:
I updated the code, I also made the String Nullable to make sure it will work with any future implementation of the webview

bugnano commented on lib/src/chatbox.dart:
```diff
@@ -52,6 +53,30 @@ class MessageActionEvent {
+    this.isValidUrl,
+  ) ;
+
+  factory UrlNavigationRequest(String url){
```
We should not do the String to Uri conversion in a public API.
A simple constructor with an Uri parameter is the best here.

The whole String to Uri conversion is better to be done in the private _shouldNavigateToUrl function.

bugnano reviewed (commented):
Thank you for the update, we're close to merging this PR, but I still ask you to do one last change.

Given that the policy at TalkJS has been to support APIs forever, and never break compatibility, we can only achieve this if we expose the bare minimum, and we expose the correct types in the first place.

bugnano commented on lib/src/chatbox.dart:
```diff
@@ -52,6 +53,16 @@ class MessageActionEvent {
+enum UrlNavigationAction { allow, deny }
+
+class UrlNavigationRequest {
+  final String? url;
```
What's the *practical* use of an UrlNavigationRequest with a null url?

Alexander-Ordina replied:
InAppWebView sometimes returns null for the url so it will not always be a String the result could be null.
I did a little more research on how InAppWebview works with urls and they do use a Uri object this means they parse the Url String to a Uri in their implementation.

This means that invalid urls won't work with there implementation, so creating a special url is only an option if it is also a valid Uri.

More information can be found here: https://inappwebview.dev/docs/webview/in-app-webview/#handle-custom-urls

I think this invalidates my previous comment about why it should be a String looking at this I do feel making it a Nullable Uri should be better. And custom urls can be handled like described in the link above. 

What do you think would be best?

Alexander-Ordina replied:
If you want we could also setup a meeting to discuss an maybe get this issue resolved a bit quicker?

bugnano replied:
I believe that a non-nullable string is the best solution here, and here's why:
While not all Strings can be valid Uris, thus resulting in a null Uri, the same is not true for Uris, which are all valid Strings, so it's better to have a non-nullable String, and then let the user decide what to do with that string.

As I see from the documentation of the InAppWebView, if an Uri is invalid, the shouldOverrideUrlLoading method is not called anyway, so it's not a problem.
And the fact that the Uri can be null in the NavigationAction.URLRequest does not affect this functionality, as I'm sure that shouldOverrideUrlLoading will not be called with a null Uri (if you read the docs, a null Uri is equivalent to "about:blank").

Alexander-Ordina replied:
Good point, I've now changed it to a non nullable String

bugnano reviewed (commented):
Thank you for the changes, I just have one question:
I appreciate very much that you got into the mentality of implementation-independence, but I don't get how a nullable string can achieve that.
Think of the user of the API:
If the url is nullable, then I expect to have an UrlNavigationRequest with a null url.
What does that mean to me, as a user?

Or, to put it another way: we decided that the url is of type String.
This means that whenever we change the implementation, we will convert whatever url type we have to String, I don't get how null can be helpful in this situation.

vickz84259 commented:
Hey @zSoNz, I'm a developer at TalkJS. Regarding the feature above, it was an intentional decision to not include it. For versions 0.6.0 and lower of our SDK, links inside the chat would open inside the webview rather than in an external browser. This was the initial motivation for implementing `navigationDelegate` as highlighted in this [issue](https://github.com/talkjs/talkjs-flutter/issues/14). This behaviour has been changed since version 0.7.0. 

Was there another use case you intended to use `navigationDelegate` for?
